Cannot find module 'react-textarea-autosize' or its corresponding type declarations

**Describe the bug**
While trying to upgrade our typescript project to storybook 6.5.x (specifically, 6.5.9), we're getting the below error when running `yarn tsc`, this was not an issue with our previous version, 6.4.0.  It looks like `@storybook/components` should have a dependency on `react-textarea-autosize` instead a `devDependency` [here](https://github.com/storybookjs/storybook/blob/v6.5.9/lib/components/package.json#L71) since it's used [here](https://github.com/storybookjs/storybook/blob/v6.5.9/lib/components/src/form/input/input.tsx#L143-L146) and available to users of storybook through `Form.Textarea`.

> node_modules/@storybook/components/dist/ts3.9/index.d.ts(278,72): error TS2307: Cannot find module 'react-textarea-autosize' or its corresponding type declarations.
> node_modules/@storybook/components/dist/ts3.9/index.d.ts(278,213): error TS2307: Cannot find module 'react-textarea-autosize' or its corresponding type declarations.

**To Reproduce**
- Update a typescript project storybook to 6.5.9
- run `yarn tsc`/`npx tsc`

**System**
  System:
    OS: macOS 12.4
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  Binaries:
    Node: 14.19.0 - ~/.nvm/versions/node/v14.19.0/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v14.13.1/bin/yarn
    npm: 6.14.16 - ~/.nvm/versions/node/v14.19.0/bin/npm
  Browsers:
    Chrome: 103.0.5060.114
    Firefox: 100.0.2
    Safari: 15.5
  npmPackages:
    @storybook/addon-a11y: ~6.5.9 => 6.5.9 
    @storybook/addon-actions: ~6.5.9 => 6.5.9 
    @storybook/addon-docs: ~6.5.9 => 6.5.9 
    @storybook/addon-essentials: ~6.5.9 => 6.5.9 
    @storybook/addon-links: ~6.5.9 => 6.5.9 
    @storybook/addon-storysource: ~6.5.9 => 6.5.9 
    @storybook/addons: ~6.5.9 => 6.5.9 
    @storybook/components: ~6.5.9 => 6.5.9 
    @storybook/react: ^6.5.9 => 6.5.9 
    @storybook/source-loader: ~6.5.9 => 6.5.9 
    @storybook/theming: ~6.5.9 => 6.5.9 

**Additional context**
Add any other context about the problem here.

--- a/code/ui/components/package.json
+++ b/code/ui/components/package.json
@@ -67,7 +67,6 @@
     "@storybook/theming": "workspace:*",
     "@storybook/types": "workspace:*",
     "memoizerific": "^1.11.3",
-    "use-resize-observer": "^9.1.0",
     "util-deprecate": "^1.0.2"
   },
   "devDependencies": {
@@ -82,7 +81,8 @@
     "react-syntax-highlighter": "^15.4.5",
     "react-textarea-autosize": "^8.3.0",
     "ts-dedent": "^2.0.0",
-    "typescript": "^5.3.2"
+    "typescript": "^5.3.2",
+    "use-resize-observer": "^9.1.0"
   },
   "peerDependencies": {
     "react": "^16.8.0 || ^17.0.0 || ^18.0.0",
@@ -94,8 +94,7 @@
   "bundler": {
     "entries": [
       "./src/index.ts"
-    ],
-    "platform": "neutral"
+    ]
   },
   "gitHead": "e6a7fd8a655c69780bc20b9749c2699e44beae17"
 }

--- a/code/ui/components/src/components/form/input/input.tsx
+++ b/code/ui/components/src/components/form/input/input.tsx
@@ -3,8 +3,27 @@ import React, { forwardRef } from 'react';
 import type { Theme, CSSObject } from '@storybook/theming';
 import { styled } from '@storybook/theming';
 
-import type { TextareaAutosizeProps } from 'react-textarea-autosize';
 import TextareaAutoResize from 'react-textarea-autosize';
+
+/**
+ * these types are copied from `react-textarea-autosize`.
+ * I copied them because of https://github.com/storybookjs/storybook/issues/18734
+ * Maybe there's some bug in `tsup` or `react-textarea-autosize`?
+ */
+type TextareaPropsRaw = React.TextareaHTMLAttributes<HTMLTextAreaElement>;
+type Style = Omit<NonNullable<TextareaPropsRaw['style']>, 'maxHeight' | 'minHeight'> & {
+  height?: number;
+};
+type TextareaHeightChangeMeta = {
+  rowHeight: number;
+};
+interface TextareaAutosizeProps extends Omit<TextareaPropsRaw, 'style'> {
+  maxRows?: number;
+  minRows?: number;
+  onHeightChange?: (height: number, meta: TextareaHeightChangeMeta) => void;
+  cacheMeasurements?: boolean;
+  style?: Style;
+}
 
 const styleResets: CSSObject = {
   // resets

--- a/code/ui/components/src/components/form/input/input.tsx
+++ b/code/ui/components/src/components/form/input/input.tsx
@@ -208,7 +208,7 @@ type TextareaProps = Omit<
   align?: Alignments;
   valid?: ValidationStates;
   height?: number;
-};
+} & React.RefAttributes<HTMLTextAreaElement>;
 export const Textarea: FC<TextareaProps> = Object.assign(
   styled(
     forwardRef<any, TextareaProps>(function Textarea({ size, valid, align, ...props }, ref) {

--- a/code/ui/components/src/components/form/input/input.tsx
+++ b/code/ui/components/src/components/form/input/input.tsx
@@ -1,4 +1,4 @@
-import type { HTMLProps, SelectHTMLAttributes } from 'react';
+import type { FC, HTMLProps, SelectHTMLAttributes } from 'react';
 import React, { forwardRef } from 'react';
 import type { Theme, CSSObject } from '@storybook/theming';
 import { styled } from '@storybook/theming';
@@ -17,7 +17,7 @@ type Style = Omit<NonNullable<TextareaPropsRaw['style']>, 'maxHeight' | 'minHeig
 type TextareaHeightChangeMeta = {
   rowHeight: number;
 };
-interface TextareaAutosizeProps extends Omit<TextareaPropsRaw, 'style'> {
+export interface TextareaAutosizeProps extends Omit<TextareaPropsRaw, 'style'> {
   maxRows?: number;
   minRows?: number;
   onHeightChange?: (height: number, meta: TextareaHeightChangeMeta) => void;
@@ -209,7 +209,7 @@ type TextareaProps = Omit<
   valid?: ValidationStates;
   height?: number;
 };
-export const Textarea = Object.assign(
+export const Textarea: FC<TextareaProps> = Object.assign(
   styled(
     forwardRef<any, TextareaProps>(function Textarea({ size, valid, align, ...props }, ref) {
       return <TextareaAutoResize {...props} ref={ref} />;

